In [23]:
import pandas as pd
import numpy as np
from pathlib import Path


In [ ]:
DATA_DIR = Path("../data/raw/MachineLearningCVE")

csv_files = sorted(DATA_DIR.glob("*.csv"))  # .glob( "*.csv ")  it mean  the last  namwe file select froim the folder

print(f"Found {len(csv_files)} CSV files")

Found 8 CSV files


In [25]:
for file in csv_files:
    print(file.name)

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Friday-WorkingHours-Morning.pcap_ISCX.csv
Monday-WorkingHours.pcap_ISCX.csv
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Tuesday-WorkingHours.pcap_ISCX.csv
Wednesday-workingHours.pcap_ISCX.csv


In [26]:
monday_file = DATA_DIR/"Monday-WorkingHours.pcap_ISCX.csv"
df = pd.read_csv(monday_file)
print("Shape: ", df.shape)

Shape:  (529918, 79)


In [27]:
df.head() # show the firs 5 flow record|

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,49188,4,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,49486,3,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


In [28]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 529918 entries, 0 to 529917
Data columns (total 79 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0    Destination Port             529918 non-null  int64  
 1    Flow Duration                529918 non-null  int64  
 2    Total Fwd Packets            529918 non-null  int64  
 3    Total Backward Packets       529918 non-null  int64  
 4   Total Length of Fwd Packets   529918 non-null  int64  
 5    Total Length of Bwd Packets  529918 non-null  int64  
 6    Fwd Packet Length Max        529918 non-null  int64  
 7    Fwd Packet Length Min        529918 non-null  int64  
 8    Fwd Packet Length Mean       529918 non-null  float64
 9    Fwd Packet Length Std        529918 non-null  float64
 10  Bwd Packet Length Max         529918 non-null  int64  
 11   Bwd Packet Length Min        529918 non-null  int64  
 12   Bwd Packet Length Mean       529918 non-null  float64


In [29]:
df.dtypes

 Destination Port                int64
 Flow Duration                   int64
 Total Fwd Packets               int64
 Total Backward Packets          int64
Total Length of Fwd Packets      int64
                                ...   
Idle Mean                      float64
 Idle Std                      float64
 Idle Max                        int64
 Idle Min                        int64
 Label                             str
Length: 79, dtype: object

In [30]:
len(df.columns)

79

In [31]:
df.columns.tolist()

[' Destination Port',
 ' Flow Duration',
 ' Total Fwd Packets',
 ' Total Backward Packets',
 'Total Length of Fwd Packets',
 ' Total Length of Bwd Packets',
 ' Fwd Packet Length Max',
 ' Fwd Packet Length Min',
 ' Fwd Packet Length Mean',
 ' Fwd Packet Length Std',
 'Bwd Packet Length Max',
 ' Bwd Packet Length Min',
 ' Bwd Packet Length Mean',
 ' Bwd Packet Length Std',
 'Flow Bytes/s',
 ' Flow Packets/s',
 ' Flow IAT Mean',
 ' Flow IAT Std',
 ' Flow IAT Max',
 ' Flow IAT Min',
 'Fwd IAT Total',
 ' Fwd IAT Mean',
 ' Fwd IAT Std',
 ' Fwd IAT Max',
 ' Fwd IAT Min',
 'Bwd IAT Total',
 ' Bwd IAT Mean',
 ' Bwd IAT Std',
 ' Bwd IAT Max',
 ' Bwd IAT Min',
 'Fwd PSH Flags',
 ' Bwd PSH Flags',
 ' Fwd URG Flags',
 ' Bwd URG Flags',
 ' Fwd Header Length',
 ' Bwd Header Length',
 'Fwd Packets/s',
 ' Bwd Packets/s',
 ' Min Packet Length',
 ' Max Packet Length',
 ' Packet Length Mean',
 ' Packet Length Std',
 ' Packet Length Variance',
 'FIN Flag Count',
 ' SYN Flag Count',
 ' RST Flag Count',
 ' P

In [36]:
df.columns = df.columns.str.strip()  # column name leading the white space

In [37]:
print(df.columns[-5:])

Index(['Idle Mean', 'Idle Std', 'Idle Max', 'Idle Min', 'Label'], dtype='str')


In [38]:
df["Label"].value_counts()

Label
BENIGN    529918
Name: count, dtype: int64

In [40]:
df["Label"].value_counts(normalize=True) * 100

Label
BENIGN    100.0
Name: proportion, dtype: float64

In [44]:
df['Label']

0         BENIGN
1         BENIGN
2         BENIGN
3         BENIGN
4         BENIGN
           ...  
529913    BENIGN
529914    BENIGN
529915    BENIGN
529916    BENIGN
529917    BENIGN
Name: Label, Length: 529918, dtype: str

In [45]:
# Show the labels clearly
print(df["Label"].unique())

<StringArray>
['BENIGN']
Length: 1, dtype: str


In [46]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/raw/MachineLearningCVE")

for file in sorted(DATA_DIR.glob("*.csv")):
    temp = pd.read_csv(file)

    temp.columns = temp.columns.str.strip()

    print("\n" + "=" * 70)
    print(file.name)
    print("Rows:", len(temp))
    print(temp["Label"].value_counts())


Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Rows: 225745
Label
DDoS      128027
BENIGN     97718
Name: count, dtype: int64

Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Rows: 286467
Label
PortScan    158930
BENIGN      127537
Name: count, dtype: int64

Friday-WorkingHours-Morning.pcap_ISCX.csv
Rows: 191033
Label
BENIGN    189067
Bot         1966
Name: count, dtype: int64

Monday-WorkingHours.pcap_ISCX.csv
Rows: 529918
Label
BENIGN    529918
Name: count, dtype: int64

Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Rows: 288602
Label
BENIGN          288566
Infiltration        36
Name: count, dtype: int64

Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Rows: 170366
Label
BENIGN                        168186
Web Attack � Brute Force        1507
Web Attack � XSS                 652
Web Attack � Sql Injection        21
Name: count, dtype: int64

Tuesday-WorkingHours.pcap_ISCX.csv
Rows: 445909
Label
BENIGN         432074
FTP-Patator      7938
SSH-Patator    

| File               |   Main traffic | Attack classes                                           |
| ------------------ | -------------: | -------------------------------------------------------- |
| Monday             | 529,918 BENIGN | None                                                     |
| Tuesday            | 432,074 BENIGN | FTP-Patator, SSH-Patator                                 |
| Wednesday          | 440,031 BENIGN | DoS Hulk, GoldenEye, slowloris, Slowhttptest, Heartbleed |
| Thursday Morning   | 168,186 BENIGN | Web Attack Brute Force, XSS, SQL Injection               |
| Thursday Afternoon | 288,566 BENIGN | Infiltration                                             |
| Friday Morning     | 189,067 BENIGN | Bot                                                      |
| Friday Afternoon   |  97,718 BENIGN | DDoS                                                     |
| Friday Afternoon   | 127,537 BENIGN | PortScan                                                 |


In [57]:
for file in sorted(DATA_DIR.glob("*.csv")):

    temp = pd.read_csv(file)
    temp.columns = temp.columns.str.strip()

    missing = temp.isnull().sum()

    print("=" * 70)
    print(file.name)

    if missing.sum() == 0:
        print("No missing values")
    else:
        print(missing[missing > 0])

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv
Flow Bytes/s    4
dtype: int64
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv
Flow Bytes/s    15
dtype: int64
Friday-WorkingHours-Morning.pcap_ISCX.csv
Flow Bytes/s    28
dtype: int64
Monday-WorkingHours.pcap_ISCX.csv
Flow Bytes/s    64
dtype: int64
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv
Flow Bytes/s    18
dtype: int64
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv
Flow Bytes/s    20
dtype: int64
Tuesday-WorkingHours.pcap_ISCX.csv
Flow Bytes/s    201
dtype: int64
Wednesday-workingHours.pcap_ISCX.csv
Flow Bytes/s    1008
dtype: int64


In [59]:
# infinite valuesd

for file in sorted(DATA_DIR.glob("*.csv")):

    temp = pd.read_csv(file)
    temp.columns = temp.columns.str.strip()

    numeric = temp.select_dtypes(include=np.number)

    inf_count = np.isinf(numeric).sum().sum()

    print(f"{file.name}: {inf_count} infinite values")

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv: 64 infinite values
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv: 727 infinite values
Friday-WorkingHours-Morning.pcap_ISCX.csv: 216 infinite values
Monday-WorkingHours.pcap_ISCX.csv: 810 infinite values
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv: 396 infinite values
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv: 250 infinite values
Tuesday-WorkingHours.pcap_ISCX.csv: 327 infinite values
Wednesday-workingHours.pcap_ISCX.csv: 1586 infinite values


In [60]:
for file in sorted(DATA_DIR.glob("*.csv")):

    temp = pd.read_csv(file)

    duplicate_count = temp.duplicated().sum()

    print(f"{file.name}: {duplicate_count} duplicate rows")

Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv: 2633 duplicate rows
Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv: 72353 duplicate rows
Friday-WorkingHours-Morning.pcap_ISCX.csv: 6888 duplicate rows
Monday-WorkingHours.pcap_ISCX.csv: 26935 duplicate rows
Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv: 35630 duplicate rows
Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv: 6066 duplicate rows
Tuesday-WorkingHours.pcap_ISCX.csv: 24065 duplicate rows
Wednesday-workingHours.pcap_ISCX.csv: 81909 duplicate rows


In [61]:
duplicate_rows = df[df.duplicated(keep=False)]

print("Duplicate rows:", len(duplicate_rows))

print(
    duplicate_rows.groupby(
        [col for col in df.columns if col != "Label"]
    )["Label"]
    .nunique()
    .value_counts()
)

Duplicate rows: 36873
Label
1    9933
Name: count, dtype: int64


In [63]:
print(duplicate_rows["Label"].unique())

<StringArray>
['BENIGN']
Length: 1, dtype: str


In [64]:
print(df["Label"].unique())

<StringArray>
['BENIGN']
Length: 1, dtype: str


In [65]:
# Reload Monday dataset from the raw CSV
monday_file = DATA_DIR / "Monday-WorkingHours.pcap_ISCX.csv"

df = pd.read_csv(monday_file)

# Normalize column names in memory
df.columns = df.columns.str.strip()

print("Shape:", df.shape)
print("Labels:")
print(df["Label"].value_counts())

Shape: (529918, 79)
Labels:
Label
BENIGN    529918
Name: count, dtype: int64


In [66]:
duplicate_rows = df[df.duplicated(keep=False)]

print("Duplicate rows:", len(duplicate_rows))
print("\nLabels among duplicate rows:")
print(duplicate_rows["Label"].value_counts())

Duplicate rows: 36873

Labels among duplicate rows:
Label
BENIGN    36873
Name: count, dtype: int64


## Conclusion

The initial analysis of the CICIDS2017 dataset provided a clear understanding of its structure, traffic classes, and data-quality characteristics.

The dataset contains multiple types of network traffic, including BENIGN traffic and several attack categories such as DDoS, PortScan, Bot, Infiltration, Web Attacks, Patator attacks, DoS attacks, and Heartbleed. The analysis also showed significant class imbalance, with some attack classes containing a large number of samples while others are extremely rare.

Several data-quality issues were identified. Missing values were found in the `Flow Bytes/s` feature, and infinite values were present in multiple files. A considerable number of duplicate rows were also observed. These findings indicate that preprocessing is required before the dataset can be used for machine-learning model development.

The analysis also showed that the original dataset should be preserved and that preprocessing should be performed separately without modifying the raw CSV files. Duplicate handling, missing-value treatment, infinite-value handling, and a leakage-resistant train/test strategy must be carefully designed.

Based on these findings, the CICIDS2017 dataset is suitable for the next stage of CyberShield AI development. The next phase will focus on preprocessing, feature preparation, and the development of a multiclass intrusion detection model.

### Key Takeaways

- The dataset contains both normal and multiple attack categories.
- The dataset has significant class imbalance.
- `Flow Bytes/s` contains missing values.
- Infinite values occur in several flow-derived features.
- Duplicate rows are present and require careful treatment.
- Raw dataset files should remain unchanged.
- Data leakage must be considered when designing the train/test strategy.
- Further preprocessing is required before model training.

**Next Phase:** `02_CICIDS2017_Preprocessing.ipynb`?